File: main/domain/include/algorithm/DepthFirstSearch.java
package algorithm;

import java.util.*;
import java.util.Map.Entry;

import comm.Define;

import domain.TownsMap;

public class DepthFirstSearch {
    private TownsMap map;

    public DepthFirstSearch() {}

    public DepthFirstSearch(TownsMap map) {
        setMap(map);
    }

    public int searchDepth(int src, int dest, int min, int max) {
        return countStopAgreed(src, dest, min, max, 0);
    }

    public int searchDistance(int src, int dest, int min, int max) {
        return countDistanceAgreed(src, dest, min, max, 0);
    }

    private int countStopAgreed(
            int current, int dest, int minUnit, int maxUnit, int currentUnit) {
        int count = 0;
        if (current == dest && currentUnit >= minUnit && currentUnit <= maxUnit)
            count += 1;
        if (currentUnit > maxUnit)
            return 0;

        Map<Integer, Integer> adjacent = getMap().getAdjacentOf(current);
        for(Entry<Integer, Integer> entry : adjacent.entrySet()) {
            count += countStopAgreed(entry.getKey(), dest, minUnit, maxUnit, currentUnit + 1);
        }
        return count;
            }

    private int countDistanceAgreed(
            int current, int dest, int minDistance, int maxDistance, int currentDistance) {
        int count = 0;
        if (current == dest && currentDistance >= minDistance && currentDistance <= maxDistance)
            count += 1;
        if (currentDistance > maxDistance)
            return 0;

        Map<Integer, Integer> adjacent = getMap().getAdjacentOf(current);
        for(Entry<Integer, Integer> entry : adjacent.entrySet()) {
            count += countDistanceAgreed(
                    entry.getKey(), dest, minDistance, maxDistance, currentDistance + entry.getValue());
        }
        return count;
            }

    private TownsMap getMap() {
        if (this.map == null) {
            this.map = new TownsMap();
        }
        return map;
    }

    public void setMap(TownsMap map) {
        this.map = map;
    }
}

File: main/domain/src/algorithm/depth_first_search.cpp
// Copyright (c) fangguojian. All rights reserved.
// author: fangguojian (destinydesigner@qq.com)
// date: 2014-08-21
// summary: depth_first_search.cpp
#include "comm/log.h"
#include "model/route.h"
#include "model/route_list.h"
#include "algorithm/depth_first_search.h"

using assignment::trains::comm::CLog;
using assignment::trains::domain::CRoute;
using assignment::trains::domain::CRouteList;
using assignment::trains::domain::CRouteIterator;
using assignment::trains::domain::CDepthFirstSearch;

CDepthFirstSearch::CDepthFirstSearch(const ITownsMap& map):
    m_pTownMap(&map)
{
}

void CDepthFirstSearch::setMap(const ITownsMap& map)
{
    m_pTownMap = &map;
}

long CDepthFirstSearch::countPath(long src, long dest, long min, long max)
{
    return countAgreedPath(src, dest, min, max, 0);
}

long CDepthFirstSearch::countAgreedPath(
    long current, long dest, long minUnit, long maxUnit, long currentUnit)
{
    long count = 0;
    MY_LOG->log(
        "countAgreedPath. current: %lu, dest: %lu,"
        " min: %lu, max: %lu, currentUnit: %lu\n",
        current, dest, minUnit, maxUnit, currentUnit);
    if (current == dest && currentUnit >= minUnit && currentUnit <= maxUnit)
        count += 1;
    if (currentUnit > maxUnit)
        return 0;

    CRouteIterator
        iter = m_pTownMap->getAllAdjacentRoutesBySource(current).getIterator();
    while(iter.hasNext())
    {
        CRoute oRoute = iter.next();
        MY_LOG->log(
            "[%lu, %lu] => %lu\n",
            oRoute.getSource().id(), oRoute.getDestination().id(),
            oRoute.getLength());
        count += countAgreedPath(
            oRoute.getDestination().id(), dest,
            minUnit, maxUnit,
            currentUnit + oRoute.getLength());
    }
    return count;
}
